File: internal/devschema/devschema_test.go
package devschema

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func writeGitDirHead(t *testing.T, gitDir, head string) {
	t.Helper()
	if err := os.MkdirAll(gitDir, 0o755); err != nil {
		t.Fatalf("mkdir git dir: %v", err)
	}
	if err := os.WriteFile(filepath.Join(gitDir, "HEAD"), []byte(head), 0o644); err != nil {
		t.Fatalf("write HEAD: %v", err)
	}
}

func writeGitHead(t *testing.T, root, branch string) {
	t.Helper()
	writeGitDirHead(t, filepath.Join(root, ".git"), "ref: refs/heads/"+branch)
}

func writeDetachedHead(t *testing.T, root string) {
	t.Helper()
	writeGitDirHead(t, filepath.Join(root, ".git"), "0123456789abcdef")
}

func writeWorktreeGitHead(t *testing.T, root, branch string) {
	t.Helper()
	gitDir := filepath.Join(root, ".gitdir")
	writeGitDirHead(t, gitDir, "ref: refs/heads/"+branch)
	if err := os.WriteFile(filepath.Join(root, ".git"), []byte("gitdir: .gitdir"), 0o644); err != nil {
		t.Fatalf("write gitdir file: %v", err)
	}
}

func TestResolveEnabledFalseOverridesSchemaName(t *testing.T) {
	t.Setenv("NODE_ENV", "development")
	t.Setenv("ENT_DEV_SCHEMA_ENABLED", "")
	cfg := &Config{
		Enabled:    false,
		SchemaName: "explicit_schema",
	}
	res, err := Resolve(cfg, Options{})
	if err != nil {
		t.Fatalf("resolve error: %v", err)
	}
	if res.Enabled {
		t.Fatalf("expected disabled when enabled=false, got enabled")
	}
}

func TestResolveIgnoreBranches(t *testing.T) {
	t.Setenv("NODE_ENV", "development")
	t.Setenv("ENT_DEV_SCHEMA_ENABLED", "")
	root := t.TempDir()
	writeGitHead(t, root, "main")
	cfg := &Config{
		Enabled:        true,
		IgnoreBranches: []string{"main"},
	}
	res, err := Resolve(cfg, Options{RepoRoot: root})
	if err != nil {
		t.Fatalf("resolve error: %v", err)
	}
	if res.Enabled {
		t.Fatalf("expected disabled for ignored branch")
	}
}

func TestResolveIgnoreBranchesEnvOverride(t *testing.T) {
	t.Setenv("NODE_ENV", "development")
	root := t.TempDir()
	writeGitHead(t, root, "main")
	t.Setenv("ENT_DEV_SCHEMA_ENABLED", "true")
	cfg := &Config{
		Enabled:        true,
		IgnoreBranches: []string{"main"},
	}
	res, err := Resolve(cfg, Options{RepoRoot: root})
	if err != nil {
		t.Fatalf("resolve error: %v", err)
	}
	if !res.Enabled {
		t.Fatalf("expected enabled when env override is true")
	}
}

func TestResolveIgnoreBranchesFromState(t *testing.T) {
	t.Setenv("NODE_ENV", "development")
	t.Setenv("ENT_DEV_SCHEMA_ENABLED", "")
	root := t.TempDir()
	writeGitHead(t, root, "main")
	res := &Result{
		Enabled:        true,
		SchemaName:     "ent_dev_main_abcd1234",
		BranchName:     "main",
		IgnoreBranches: []string{"main"},
	}
	if err := WriteState(res, Options{RepoRoot: root}); err != nil {
		t.Fatalf("write state: %v", err)
	}
	out, err := Resolve(nil, Options{RepoRoot: root})
	if err != nil {
		t.Fatalf("resolve error: %v", err)
	}
	if out.Enabled {
		t.Fatalf("expected disabled via state ignoreBranches")
	}
}

func TestResolveSanitizesExplicitSchemaName(t *testing.T) {
	t.Setenv("NODE_ENV", "development")
	t.Setenv("ENT_DEV_SCHEMA_ENABLED", "")
	cfg := &Config{
		Enabled:    true,
		SchemaName: "123bad",
	}
	res, err := Resolve(cfg, Options{})
	if err != nil {
		t.Fatalf("resolve error: %v", err)
	}
	if res.SchemaName != "schema_123bad" {
		t.Fatalf("expected sanitized schema name, got %q", res.SchemaName)
	}
}

func TestResolveRuntimeConfigDerivesBranchSchemaWithoutState(t *testing.T) {
	t.Setenv("NODE_ENV", "development")
	t.Setenv("ENT_DEV_SCHEMA_ENABLED", "")
	root := t.TempDir()
	writeGitHead(t, root, "feature/add-dev-schema")

	res, err := Resolve(&Config{
		Enabled:       true,
		IncludePublic: true,
	}, Options{RepoRoot: root})
	if err != nil {
		t.Fatalf("resolve error: %v", err)
	}
	if !res.Enabled {
		t.Fatalf("expected enabled")
	}
	if res.BranchName != "feature/add-dev-schema" {
		t.Fatalf("expected branch name, got %q", res.BranchName)
	}
	if !res.IncludePublic {
		t.Fatalf("expected includePublic to be preserved")
	}
	expectedPrefix := sanitizeIdentifier(DefaultPrefix) + "_" + slugify("feature/add-dev-schema") + "_"
	if !strings.HasPrefix(res.SchemaName, expectedPrefix) {
		t.Fatalf("expected schema name prefix %q, got %q", expectedPrefix, res.SchemaName)
	}
}

func TestResolveStateBranchMismatchFailsClosed(t *testing.T) {
	t.Setenv("NODE_ENV", "development")
	t.Setenv("ENT_DEV_SCHEMA_ENABLED", "")
	root := t.TempDir()
	writeGitHead(t, root, "main")
	if err := WriteState(&Result{
		Enabled:    true,
		SchemaName: "ent_dev_feature_abcd1234",
		BranchName: "feature",
	}, Options{RepoRoot: root}); err != nil {
		t.Fatalf("write state: %v", err)
	}

	_, err := Resolve(nil, Options{RepoRoot: root})
	if err == nil {
		t.Fatalf("expected branch mismatch error")
	}
	if !strings.Contains(err.Error(), `generated for branch "feature" but current branch is "main"`) {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestResolveStateUnknownCurrentBranchFailsClosed(t *testing.T) {
	t.Setenv("NODE_ENV", "development")
	t.Setenv("ENT_DEV_SCHEMA_ENABLED", "")
	root := t.TempDir()
	writeDetachedHead(t, root)
	if err := WriteState(&Result{
		Enabled:    true,
		SchemaName: "ent_dev_feature_abcd1234",
		BranchName: "feature",
	}, Options{RepoRoot: root}); err != nil {
		t.Fatalf("write state: %v", err)
	}

	_, err := Resolve(nil, Options{RepoRoot: root})
	if err == nil {
		t.Fatalf("expected unknown branch error")
	}
	if !strings.Contains(err.Error(), "could not be determined") {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestResolveGitBranchFromWorktreeGitFile(t *testing.T) {
	t.Setenv("NODE_ENV", "development")
	t.Setenv("ENT_DEV_SCHEMA_ENABLED", "")
	root := t.TempDir()
	writeWorktreeGitHead(t, root, "feature/worktree")

	res, err := Resolve(&Config{Enabled: true}, Options{RepoRoot: root})
	if err != nil {
		t.Fatalf("resolve error: %v", err)
	}
	if res.BranchName != "feature/worktree" {
		t.Fatalf("expected worktree branch, got %q", res.BranchName)
	}
}

func TestBuildSchemaNameLength(t *testing.T) {
	t.Setenv("NODE_ENV", "development")
	branch := strings.Repeat("feature-super-long-", 10)
	name := buildSchemaName(branch)
	if len(name) > MaxSchemaLen {
		t.Fatalf("expected schema name <= %d, got %d", MaxSchemaLen, len(name))
	}
	if !strings.HasPrefix(name, sanitizeIdentifier(DefaultPrefix)+"_") {
		t.Fatalf("expected prefix %q, got %q", sanitizeIdentifier(DefaultPrefix), name)
	}
}

File: ts/src/core/dev_schema.test.ts
import fs from "fs";
import os from "os";
import path from "path";
import { isDevSchemaEnabled, resolveDevSchema } from "./dev_schema";

function writeHead(gitDir: string, head: string) {
  fs.mkdirSync(gitDir, { recursive: true });
  fs.writeFileSync(path.join(gitDir, "HEAD"), head, "utf8");
}

function makeRepo(branch: string): string {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), "ent-devschema-"));
  writeHead(path.join(dir, ".git"), `ref: refs/heads/${branch}`);
  return dir;
}

function makeDetachedRepo(): string {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), "ent-devschema-"));
  writeHead(path.join(dir, ".git"), "0123456789abcdef");
  return dir;
}

function makeWorktreeRepo(branch: string): string {
  const dir = fs.mkdtempSync(path.join(os.tmpdir(), "ent-devschema-"));
  writeHead(path.join(dir, ".gitdir"), `ref: refs/heads/${branch}`);
  fs.writeFileSync(path.join(dir, ".git"), "gitdir: .gitdir", "utf8");
  return dir;
}

function writeState(
  repo: string,
  state: {
    schemaName: string;
    branchName?: string;
    includePublic?: boolean;
    ignoreBranches?: string[];
  },
) {
  const stateDir = path.join(repo, "src", "schema", ".ent");
  fs.mkdirSync(stateDir, { recursive: true });
  fs.writeFileSync(
    path.join(stateDir, "dev_schema.json"),
    JSON.stringify(state),
    "utf8",
  );
}

describe("dev schema enablement", () => {
  const originalEnv = { ...process.env };
  const originalCwd = process.cwd();

  afterEach(() => {
    for (const key of Object.keys(process.env)) {
      if (!(key in originalEnv)) {
        delete process.env[key];
      }
    }
    for (const [key, value] of Object.entries(originalEnv)) {
      if (value === undefined) {
        delete process.env[key];
      } else {
        process.env[key] = value;
      }
    }
    process.chdir(originalCwd);
  });

  test("enabled false overrides schemaName", () => {
    const repo = makeRepo("main");
    process.chdir(repo);
    process.env.NODE_ENV = "development";
    delete process.env.ENT_DEV_SCHEMA_ENABLED;

    expect(isDevSchemaEnabled({ enabled: false, schemaName: "explicit" })).toBe(
      false,
    );
  });

  test("ignoreBranches disables on matching branch", () => {
    const repo = makeRepo("main");
    process.chdir(repo);
    process.env.NODE_ENV = "development";
    delete process.env.ENT_DEV_SCHEMA_ENABLED;

    expect(
      isDevSchemaEnabled({ enabled: true, ignoreBranches: ["main"] }),
    ).toBe(false);
  });

  test("env override enables even when ignored", () => {
    const repo = makeRepo("main");
    process.chdir(repo);
    process.env.NODE_ENV = "development";
    process.env.ENT_DEV_SCHEMA_ENABLED = "true";

    expect(
      isDevSchemaEnabled({ enabled: true, ignoreBranches: ["main"] }),
    ).toBe(true);
  });

  test("state ignoreBranches disables when no config provided", () => {
    const repo = makeRepo("main");
    process.chdir(repo);
    process.env.NODE_ENV = "development";
    delete process.env.ENT_DEV_SCHEMA_ENABLED;
    writeState(repo, {
      schemaName: "ent_dev_main_abcd1234",
      branchName: "main",
      ignoreBranches: ["main"],
    });

    expect(isDevSchemaEnabled()).toBe(false);
  });
});

describe("dev schema resolution", () => {
  const originalEnv = { ...process.env };
  const originalCwd = process.cwd();

  afterEach(() => {
    for (const key of Object.keys(process.env)) {
      if (!(key in originalEnv)) {
        delete process.env[key];
      }
    }
    for (const [key, value] of Object.entries(originalEnv)) {
      if (value === undefined) {
        delete process.env[key];
      } else {
        process.env[key] = value;
      }
    }
    process.chdir(originalCwd);
  });

  test("sanitizes explicit schemaName", () => {
    const repo = makeRepo("main");
    process.chdir(repo);
    process.env.NODE_ENV = "development";

    const res = resolveDevSchema({ enabled: true, schemaName: "123Bad" });
    expect(res.schemaName).toBe("schema_123bad");
  });

  test("runtime config derives schema without state", () => {
    const repo = makeRepo("feature/add-dev-schema");
    process.chdir(repo);
    process.env.NODE_ENV = "development";
    delete process.env.ENT_DEV_SCHEMA_ENABLED;

    const res = resolveDevSchema({ enabled: true, includePublic: true });
    expect(res.enabled).toBe(true);
    expect(res.branchName).toBe("feature/add-dev-schema");
    expect(res.includePublic).toBe(true);
    expect(res.schemaName).toMatch(
      /^ent_dev_feature_add_dev_schema_[0-9a-f]{8}$/,
    );
  });

  test("runtime config ignores stale state file and uses runtime settings", () => {
    const repo = makeRepo("main");
    process.chdir(repo);
    process.env.NODE_ENV = "development";
    delete process.env.ENT_DEV_SCHEMA_ENABLED;
    writeState(repo, {
      schemaName: "ent_dev_feature_abcd1234",
      branchName: "feature",
      includePublic: false,
    });

    const res = resolveDevSchema({ enabled: true, includePublic: true });
    expect(res.branchName).toBe("main");
    expect(res.includePublic).toBe(true);
    expect(res.schemaName).toMatch(/^ent_dev_main_[0-9a-f]{8}$/);
  });

  test("throws on branch mismatch when using state file", () => {
    const repo = makeRepo("main");
    process.chdir(repo);
    process.env.NODE_ENV = "development";
    writeState(repo, {
      schemaName: "ent_dev_feature_abcd1234",
      branchName: "feature",
    });

    expect(() => resolveDevSchema()).toThrow(
      /dev branch schema state was generated for branch "feature" but current branch is "main"/,
    );
  });

  test("throws when current branch cannot be determined in state-file mode", () => {
    const repo = makeDetachedRepo();
    process.chdir(repo);
    process.env.NODE_ENV = "development";
    writeState(repo, {
      schemaName: "ent_dev_feature_abcd1234",
      branchName: "feature",
    });

    expect(() => resolveDevSchema()).toThrow(/could not be determined/);
  });

  test("resolves worktree gitdir files", () => {
    const repo = makeWorktreeRepo("feature/worktree");
    process.chdir(repo);
    process.env.NODE_ENV = "development";
    delete process.env.ENT_DEV_SCHEMA_ENABLED;

    const res = resolveDevSchema({ enabled: true });
    expect(res.branchName).toBe("feature/worktree");
    expect(res.schemaName).toMatch(/^ent_dev_feature_worktree_[0-9a-f]{8}$/);
  });
});
